src/mds/FSMap.cc

```diff
@@ -0,0 +1,762 @@
+           << fs->mds_map.max_mds << " up";
+    } else {
+      for (auto i : filesystems) {
+        auto fs = i.second;
+        *out << " " << fs->mds_map.fs_name << "-" << fs->mds_map.up.size() << "/"
+             << fs->mds_map.in.size() << "/" << fs->mds_map.max_mds << " up";
+      }
+    }
+  }
+
+  if (f) {
+    f->open_array_section("by_rank");
+  }
+
+  const auto all_info = get_mds_info();
+  for (const auto &p : all_info) {
+    const auto &info = p.second;
+    string s = ceph_mds_state_name(info.state);
+    if (info.laggy()) {
+      s += "(laggy or crashed)";
+    }
+
+    const fs_cluster_id_t fscid = mds_roles.at(info.global_id);
+
+    if (info.rank != MDS_RANK_NONE) {
+      if (f) {
+        f->open_object_section("mds");
+        f->dump_unsigned("filesystem_id", fscid);
+        f->dump_unsigned("rank", info.rank);
+        f->dump_string("name", info.name);
+        f->dump_string("status", s);
+        f->close_section();
+      } else {
+        by_rank[mds_role_t(fscid, info.rank)] = info.name + "=" + s;
+      }
+    } else {
+      by_state[s]++;
+    }
+  }
+
+  if (f) {
+    f->close_section();
+  } else {
+    if (!by_rank.empty()) {
+      if (filesystems.size() > 1) {
+        // Disambiguate filesystems
+        std::map<std::string, std::string> pretty;
+        for (auto i : by_rank) {
+          const auto &fs_name = filesystems.at(i.first.fscid)->mds_map.fs_name;
+          std::ostringstream o;
+          o << "[" << fs_name << ":" << i.first.rank << "]";
+          pretty[o.str()] = i.second;
+        }
+        *out << " " << pretty;
+      } else {
+        *out << " " << by_rank;
+      }
+    }
+  }
+
+  for (map<string,int>::reverse_iterator p = by_state.rbegin(); p != by_state.rend(); ++p) {
+    if (f) {
+      f->dump_unsigned(p->first.c_str(), p->second);
+    } else {
+      *out << ", " << p->second << " " << p->first;
+    }
+  }
+
+  size_t failed = 0;
+  size_t damaged = 0;
+  for (auto i : filesystems) {
+    auto fs = i.second;
+    failed += fs->mds_map.failed.size();
+    damaged += fs->mds_map.damaged.size();
+  }
+
+  if (failed > 0) {
+    if (f) {
+      f->dump_unsigned("failed", failed);
+    } else {
+      *out << ", " << failed << " failed";
+    }
+  }
+
+  if (damaged > 0) {
+    if (f) {
+      f->dump_unsigned("damaged", damaged);
+    } else {
+      *out << ", " << damaged << " damaged";
+    }
+  }
+  //if (stopped.size())
+  //out << ", " << stopped.size() << " stopped";
+}
+
+void FSMap::get_health(list<pair<health_status_t,string> >& summary,
+			list<pair<health_status_t,string> > *detail) const
+{
+  for (auto i : filesystems) {
+    auto fs = i.second;
+
+    // TODO: move get_health up into here so that we can qualify
+    // all the messages with what filesystem they're talking about
+    fs->mds_map.get_health(summary, detail);
+  }
+}
+
+void FSMap::encode(bufferlist& bl, uint64_t features) const
+{
+  ENCODE_START(10, 10, bl);
```

Comment:
Hmm, I think the 10 originally came from where ev=9 was partway through the encoded MDSMap.  But since the MDSMap's first header is at version 5, yes, we can set the FSMap's verision to 6 instead of 10.